shell/src/main/java/alluxio/shell/command/CpCommand.java

```diff
@@ -0,0 +1,228 @@
+      if (dstStatus != null && dstStatus.isFolder()) {
+        dstPath = new AlluxioURI(PathUtils.concatPath(dstPath.getPath(), srcPath.getName()));
+      }
+      copyFile(srcPath, dstPath);
+    } else {
+      if (!recursive) {
+        throw new IOException(
+            srcPath.getPath() + " is a directory, to copy it please use \"cp -R <src> <dst>\"");
+      }
+
+      List<URIStatus> statuses;
+      statuses = mFileSystem.listStatus(srcPath);
+
+      if (dstStatus != null) {
+        if (!dstStatus.isFolder()) {
+          throw new InvalidPathException(ExceptionMessage.DESTINATION_CANNOT_BE_FILE.getMessage());
+        }
+        // if copying a directory to an existing directory, the copied directory will become a
+        // subdirectory of the destination
+        if (srcStatus.isFolder()) {
+          dstPath = new AlluxioURI(PathUtils.concatPath(dstPath.getPath(), srcPath.getName()));
+          mFileSystem.createDirectory(dstPath);
+          System.out.println("Created directory: " + dstPath.getPath());
+        }
+      }
+
+      if (dstStatus == null) {
+        mFileSystem.createDirectory(dstPath);
+        System.out.println("Created directory: " + dstPath.getPath());
+      }
+
+      List<String> errorMessages = new ArrayList<>();
+      for (URIStatus status : statuses) {
+        try {
+          copy(new AlluxioURI(srcPath.getScheme(), srcPath.getAuthority(), status.getPath()),
+              new AlluxioURI(dstPath.getScheme(), dstPath.getAuthority(),
+                  PathUtils.concatPath(dstPath.getPath(), status.getName())), recursive);
+        } catch (IOException e) {
+          errorMessages.add(e.getMessage());
+        }
+      }
+
+      if (errorMessages.size() != 0) {
+        throw new IOException(Joiner.on('\n').join(errorMessages));
+      }
+    }
+  }
+
+  /**
+   * Copies a file in the Alluxio filesystem.
+   *
+   * @param srcPath the source {@link AlluxioURI} (has to be a file)
+   * @param dstPath the destination path in the Alluxio filesystem
+   * @throws AlluxioException when Alluxio exception occurs
+   * @throws IOException when non-Alluxio exception occurs
+   */
+  private void copyFile(AlluxioURI srcPath, AlluxioURI dstPath)
+      throws AlluxioException, IOException {
+    try (Closer closer = Closer.create()) {
+      OpenFileOptions openFileOptions = OpenFileOptions.defaults().setReadType(ReadType.NO_CACHE);
+      FileInStream is = closer.register(mFileSystem.openFile(srcPath, openFileOptions));
+      CreateFileOptions createFileOptions = CreateFileOptions.defaults();
+      FileOutStream os = closer.register(mFileSystem.createFile(dstPath, createFileOptions));
+      IOUtils.copy(is, os);
+      System.out.println("Copied " + srcPath + " to " + dstPath.getPath());
+    }
+  }
+
+  @Override
+  public String getUsage() {
+    return "cp [-R] <src> <dst>";
+  }
+
+  @Override
+  public String getDescription() {
+    return "Copies a file or a directory in the Alluxio filesystem."
```

Comment:
it does ... the caveat is that it is inefficient because it copies everything sequentially